vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -477,6 +482,160 @@ def query_metrics_wrapper(self, query_specs):
+
+            for cluster in new_performance_metrics:
+                for entity_type in cluster:
+                    if len(entity_type.value) > 0:
+                        resource_type = entity_type.value[0].metricId.dynamicProperty[0][0]
+                        for given_metric in entity_type.value:
+                            if self.log.isEnabledFor(logging.DEBUG):
+                                self.log.debug(
+                                    "Processing metric `%s`: resource_type=`%s`, result=`%s`",
+                                    given_metric.metricId.label,
+                                    type(given_metric),
+                                    str(given_metric).replace("\n", "\\n"),
+                                )
+                            latest_value = given_metric.values.split(',')[-1]
+                            if latest_value != 'None':
+                                if given_metric.metricId.label in VSAN_PERCENT_METRICS:
+                                    latest_value = float(latest_value) / 100.0
+
+                                hostname = None
+                                if resource_type == 'cluster':
+                                    tags = self.infrastructure_cache.get_mor_tags(
+                                        given_metric.metricId.dynamicProperty[0][1]
+                                    )
+                                    tags.extend(
+                                        self.infrastructure_cache.get_mor_props(
+                                            given_metric.metricId.dynamicProperty[0][1]
+                                        )['tags']
+                                    )
+                                    tags.extend(self._config.base_tags)
+                                elif resource_type == 'host':
+                                    tags = (
+                                        [
+                                            t
+                                            for t in self.infrastructure_cache.get_mor_tags(
+                                                given_metric.metricId.dynamicProperty[0][1]
+                                            )
+                                            if t.split(":", 1)[0] in self._config.excluded_host_tags
+                                        ]
+                                        if self._config.excluded_host_tags
+                                        else []
+                                    )
+                                    tags.extend(
+                                        self.infrastructure_cache.get_mor_props(
+                                            given_metric.metricId.dynamicProperty[0][1]
+                                        )['tags']
+                                    )
+                                    tags.extend(self._config.base_tags)
+                                    hostname = given_metric.metricId.dynamicProperty[0][2]
+
```

Comment:
can we do this once per resource rather than every metric, since we know the tags will be the same for each unique resource?
```suggestion 
                               entity = given_metric.metricId.dynamicProperty[0][1]
                               resource_tags = self.infrastructure_cache.get_mor_tags(entity)
                               tags = copy.deepcopy(resource_tags)
                               tags.extend(self.infrastructure_cache.get_mor_props(entity)['tags'])
                               if resource_type == 'host':
                                    tags = (
                                        [t for t in resource_tags if t.split(":", 1)[0] in self._config.excluded_host_tags]
                                        if self._config.excluded_host_tags
                                        else []
                                    )
                                    tags.extend(self._config.base_tags)
                                    hostname = given_metric.metricId.dynamicProperty[0][2]

```
nit for reusing some of the logic for each resource now that disks are not involved

Additionally, can we do this once per resource rather than every metric, since we know the tags will be the same for each unique resource?

Reply:
Great idea...I made your change, and I moved this code up so that it runs per-resource rather than per-metric.